contracts/multisig/src/signing.rs

```diff
@@ -26,72 +26,102 @@ impl SigningSession {
+            return Err(ContractError::NotAParticipant {
+                session_id: session.id,
+                signer,
             });
         }
+        _ => {}
+    }
 
-        match key.pub_keys.get(&signer) {
-            Some(pub_key) if !signature.verify(&self.msg, pub_key)? => {
-                return Err(ContractError::InvalidSignature {
-                    session_id: self.id,
-                    signer,
-                });
-            }
-            None => {
-                return Err(ContractError::NotAParticipant {
-                    session_id: self.id,
-                    signer,
-                });
-            }
-            _ => {}
-        }
+    Ok(())
+}
 
-        self.signatures.insert(signer, signature);
+fn signers_weight(
+    signatures: HashMap<String, Signature>,
+    snapshot: &Snapshot,
+) -> StdResult<Uint256> {
+    signatures
+        .keys()
+        .map(|addr| -> StdResult<Uint256> {
+            Ok(snapshot
+                .participants
+                .get(addr)
+                .expect("violated invariant: signature submitted by non-participant")
+                .weight
+                .into())
+        })
+        .sum()
+}
 
-        // TODO: may need to also check state != Completed if expiration is ever introduced
-        if self.signers_weight(&key.snapshot) >= key.snapshot.quorum.into() {
-            self.state = MultisigState::Completed;
-        }
+pub fn sign(
+    store: &mut dyn Storage,
+    session: &SigningSession,
+    key: &Key,
+    signer: String,
+    signature: Signature,
+) -> Result<Signature, ContractError> {
+    validate_session_signature(session, key, signer.clone(), &signature)?;
+
+    SIGNATURES.update(
+        store,
+        (session.id.u64(), &signer),
+        |sig| -> Result<Signature, ContractError> {
+            match sig {
+                Some(_) => Err(ContractError::DuplicateSignature {
+                    session_id: session.id,
+                    signer: signer.clone(),
+                }),
+                None => Ok(signature),
+            }
+        },
+    )
+}
 
-        Ok(())
-    }
+pub fn update_session_state(
+    store: &mut dyn Storage,
+    mut session: SigningSession,
+    snapshot: &Snapshot,
+) -> Result<MultisigState, ContractError> {
+    let signatures = session_signatures(store, session.id.u64())?;
 
-    fn signers_weight(&self, snapshot: &Snapshot) -> Uint256 {
-        self.signatures
-            .keys()
-            .map(|addr| -> Uint256 {
-                snapshot
-                    .participants
-                    .get(addr)
-                    .expect("violated invariant: signature submitted by non-participant")
-                    .weight
-                    .into()
-            })
-            .sum()
+    if signers_weight(signatures, snapshot)? >= snapshot.quorum.into() {
+        session.state = MultisigState::Completed;
+        SIGNING_SESSIONS.save(store, session.id.u64(), &session)?;
```

Comment:
Is this what's causing the 10k increase when the threshold is reached?

Reply:
yes, I commented out the `SIGNING_SESSIONS.save(store, session.id.u64(), &session)?` and instead of a 10k increase its only like 200. Maybe we don't really need to store the state, we could re-calculate it on the fly whenever is needed?

Reply:
calculating on the fly sounds good to me

Reply:
removed state from storage and instead recalculated everytime it is needed. Here is the new table of gas used:



signature index | gas cost | diff from previous
-- | -- | --
1 | 153127 | 0
2 | 155398 | 2271
3 | 156966 | 1568
4 (threshold reached) | 158666 | 1700